buildtest/menu/build.py

```diff
@@ -74,10 +74,65 @@ def discover_configs(config_file):
         logger.error(msg)
         sys.exit(msg)
 
+    # return all configs with absolute path
+    tmp = [os.path.abspath(config) for config in config_files]
+    config_files = tmp
+
     logger.info(f"Found the following config files: {config_files}")
     return config_files
 
 
+def exclude_configs(config_files, exclude_list):
```

Comment:
This is way overly complicated a function for such a simple action! We basically need to get the resolved paths for all discovered configs, create a single regular expression that combines the exclude patterns, and then (in one loop) go over them and keep ones not excluded. Here is an example of a function I wrote that does the same thing in just a few lines: https://github.com/urlstechie/urlchecker-python/blob/04c84629eabaa70a131a670308eb54c741d62cf2/urlchecker/core/fileproc.py#L37

And then is used in a single iterator https://github.com/urlstechie/urlchecker-python/blob/04c84629eabaa70a131a670308eb54c741d62cf2/urlchecker/core/fileproc.py#L82

Reply:
When i was implementing ``--exclude`` i didn't intend for regular expression. Regex would be nice but i foresee that to be a significant change or maybe even a new option. Right now i mimic ``-c`` with ``-x`` which supports file or directory. Both of them don't support regex. 

I think advanced users that know regex would prefer a pattern search for both ``-c`` and ``-x``, but most novice users would prefer just passing a file or directory.

Would it make sense to stand up a new option for regular expression, or we can support regex into both ``-c`` and ``-x`` to be consistent. One thing that i realized was ``-c`` can't be used twice which i can change by adding action ``append`` so something like

```
buildtest build -c dir1 -c dir2
```

Reply:
The --exclude doesn't take a regexp, it turns into a list, and that would be parsed into a regular expression. E.g.,:

```python
exclude = ["dir1", "dir2"]
```
turns into

```python
"(dir1|dir2)"
```
The command line interface you have does not change.

Reply:
I see the regex examples i am not entirely if regex will do trick or my lack of understanding. For a directory i need to retrieve all yml files and add to exclude list before checking if they are already discovered so i can remove them.

Reply:
Maybe this will help:

 1. You start with a list of files, config_files
 2. You run them through an iterate to filter that determines if they should be included
 3. The filter takes the entire list of excludes and generates _one_ regular expression to check the _one_ config file path that is being passed over
 4. The function returns True or False to determine to include or not. E.g.,

```
exclude_list = ["dir1", "dir2"]
config_files = [config for config in config_files if include_file(config, exclude_list)]
```
And the include file function is similar to what I showed you.

Reply:
Okay i fixed this now and i was able to fix resolve path to support both shell and user expansion and applied resolve path to config files too. So this two tests are now discovered properly

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c $PWD/examples
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
bzip2                          bzip_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_help                      PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip2_help                     PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 10 tests
Passed Tests: 8/10 Percentage: 80.0%
Failed Tests: 2/10 Percentage: 20.0%
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c ~/gpfs/buildtest-framework/examples
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
bzip2                          bzip_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_help                      PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip2_help                     PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 10 tests
Passed Tests: 8/10 Percentage: 80.0%
Failed Tests: 2/10 Percentage: 20.0%

```

Reply:
Thanks @vsoch  the solution does work i had to make a minor adjustment to apply ``resolve_path`` before running regex since paths needed to be resolved for regex to work, i had applied some logging to show some details.

Here is an example with user expansion
 
```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c ~/gpfs/buildtest-framework/examples -x ~/gpfs/buildtest-framework/examples/
There are no config files to process.
```

The logfile will have the following
```
2020-04-10 09:38:39,875 [build.py:155 - func_build_subcmd() ] - [DEBUG] The exclude config pattern are the following: -e ['/u/users/ssi29/gpfs/buildtest-framework/examples/']
2020-04-10 09:38:39,876 [build.py:105 - include_file() ] - [DEBUG] white list pattern before resolving paths: ['/u/users/ssi29/gpfs/buildtest-framework/examples/']
2020-04-10 09:38:39,876 [build.py:109 - include_file() ] - [DEBUG] white list pattern after resolving paths: ['/mxg-hpc/users/ssi29/buildtest-framework/examples']
2020-04-10 09:38:39,876 [build.py:112 - include_file() ] - [DEBUG] Applying Regular Expression Search: (/mxg-hpc/users/ssi29/buildtest-framework/examples) to file: /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
```

Here is an example with shell expansion where only ``examples/benchmark`` is supposed to be processed

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/ -x $HOME/gpfs/buildtest-framework/examples/script
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 2 tests
Passed Tests: 2/2 Percentage: 100.0%
Failed Tests: 0/2 Percentage: 0.0%
```